### PR TITLE
ITK 5.2 migration

### PR DESCRIPTION
**Pull request by dzenanz**
A cleaned-up version of #378. Closes #378.

### CODE REVIEW

File: src/Application/Project/Project.cc [hunk 1]
```diff
@@ -592,20 +592,20 @@ bool ProjectPrivate::initialize_session_database()
 
   // Create table for storing the database version
   sql_statements += "CREATE TABLE database_version "
-    "(version INTEGER NOT NULL PRIMARY KEY);";
+    "(version INTEGER NOT nullptr PRIMARY KEY);";
```
Comment (dcwhite) [inline src/Application/Project/Project.cc:595]: These are SQL statements so they need to be reverted back to using NULL, not nullptr. Bad "replace all" bug introduced here...

File: src/Application/Project/Project.cc [hunk 2]
```diff
@@ -592,20 +592,20 @@ bool ProjectPrivate::initialize_session_database()
...
 
   // Create table for sessions
   sql_statements += "CREATE TABLE session "
-    "(session_id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT, "
-    "session_name TEXT NOT NULL, "
+    "(session_id INTEGER NOT nullptr PRIMARY KEY AUTOINCREMENT, "
```
Comment (dcwhite) [inline src/Application/Project/Project.cc:599]: Ditto

File: src/Application/Project/Project.cc [hunk 3]
```diff
@@ -659,35 +659,35 @@ bool ProjectPrivate::initialize_provenance_database()
 
   // Create table for storing inputs of each provenance step
   sql_statements += "CREATE TABLE provenance_input "
-    "(input_id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT, "
-    "prov_step_id INTEGER NOT NULL REFERENCES provenance_step(prov_step_id) ON DELETE CASCADE, "
-    "prov_id INTEGER NOT NULL);";
+    "(input_id INTEGER NOT nullptr PRIMARY KEY AUTOINCREMENT, "
```
Comment (dcwhite) [inline src/Application/Project/Project.cc:662]: And more ditto...revert the whole file...

File: src/Application/ProjectManager/ProjectManager.cc [hunk 4]
```diff
@@ -114,14 +114,14 @@ bool ProjectManagerPrivate::initialize_project_database()
 
   // Create table for storing the database version
   sql_statements << "CREATE TABLE database_version "
-                 << "(version INTEGER NOT NULL PRIMARY KEY);";
+                 << "(version INTEGER NOT nullptr PRIMARY KEY);";
```
Comment (dcwhite) [inline src/Application/ProjectManager/ProjectManager.cc:117]: Revert this file too

File: src/ThirdParty/tinyxml/tinyxml.h [hunk 5]
```diff
@@ -276,7 +276,7 @@ class TiXmlBase
 		TIXML_ERROR_PARSING_COMMENT,
 		TIXML_ERROR_PARSING_DECLARATION,
 		TIXML_ERROR_DOCUMENT_EMPTY,
-		TIXML_ERROR_EMBEDDED_NULL,
+		TIXML_ERROR_EMBEDDED_nullptr,
```
Comment (dcwhite) [inline src/ThirdParty/tinyxml/tinyxml.h:279]:
```suggestion
		TIXML_ERROR_EMBEDDED_NULL,
```

Another mistaken replacement

File: src/ThirdParty/tinyxml/tinyxmlparser.cpp [hunk 6]
```diff
@@ -661,7 +661,7 @@ void TiXmlDocument::StreamIn( std::istream * in, TIXML_STRING * tag )
 			int c = in->get();
 			if ( c <= 0 )
 			{
-				SetError( TIXML_ERROR_EMBEDDED_NULL, 0, 0, TIXML_ENCODING_UNKNOWN );
+				SetError( TIXML_ERROR_EMBEDDED_nullptr, 0, 0, TIXML_ENCODING_UNKNOWN );
```
Comment (dcwhite) [inline src/ThirdParty/tinyxml/tinyxmlparser.cpp:664]:
```suggestion
				SetError( TIXML_ERROR_EMBEDDED_NULL, 0, 0, TIXML_ENCODING_UNKNOWN );
```

File: src/ThirdParty/tinyxml/tinyxmlparser.cpp [hunk 7]
```diff
@@ -911,7 +911,7 @@ void TiXmlElement::StreamIn (std::istream * in, TIXML_STRING * tag)
 		{
 			TiXmlDocument* document = GetDocument();
 			if ( document )
-				document->SetError( TIXML_ERROR_EMBEDDED_NULL, 0, 0, TIXML_ENCODING_UNKNOWN );
+				document->SetError( TIXML_ERROR_EMBEDDED_nullptr, 0, 0, TIXML_ENCODING_UNKNOWN );
```
Comment (dcwhite) [inline src/ThirdParty/tinyxml/tinyxmlparser.cpp:914]: Revert whole file

**Review by dcwhite — CHANGES_REQUESTED** [review]

**Comment by RubioJr9** [comment]
Built on Arch Linux w/ Qt 5.15.1

This is the error I get building with default settings.
```
In file included from /home/tpat/repos/seg3d/src/Application/LayerIO/Matlab73LayerImporter.cc:31:
/home/tpat/repos/seg3d/bin/Externals/Install/ITK_external/include/ITK-5.1/itk_H5Cpp.h:27:11: fatal error: itkhdf5/cpp/H5Cpp.h: No such file or directory
   27 | # include "itkhdf5/cpp/H5Cpp.h"
```

This is what I get if I try setting ITK_USE_SYSTEM_HDF5=ON in cmake.
```
/home/tpat/repos/seg3d/src/Application/LayerIO/Matlab73LayerImporter.cc: In member function 'bool Seg3D::Matlab73LayerImporterPrivate::scan_mat_struct(hid_t, H5::Group&, std::string&)':
/home/tpat/repos/seg3d/src/Application/LayerIO/Matlab73LayerImporter.cc:519:40: error: too few arguments to function 'herr_t H5Oget_info3(hid_t, H5O_info2_t*, unsigned int)'
  519 |       if (H5Oget_info(obj_id, &obj_info) < 0)
```

File: src/Application/DatabaseManager/DatabaseManager.cc [hunk 8]
```diff
@@ -78,7 +78,7 @@ DatabaseManager::DatabaseManager( const DatabaseManager& src ) :
     // Copy the database content from the source
     sqlite3_backup* db_backup_obj = sqlite3_backup_init( this->private_->database_,
       "main", src.private_->database_, "main" );
-    if ( db_backup_obj == nullptr )
```
Comment (dzenanz) [inline src/Application/DatabaseManager/DatabaseManager.cc]: I think that `DatabaseManager` is pure C++ file, so leaving `nullptr` in here is fine. I will edit this commit.

Comment (allywarner) [reply src/Application/DatabaseManager/DatabaseManager.cc]: Anything that has to do with SQL requires NULL and not nullptr. They need to stay NULL or there will be SQL errors.

**Comment by dzenanz** [comment]
@RubioJr9 switch from ITK 5.1 to `master` (provisionally 5.2) breaks the incremental build. You could try cleaning up only ITK-related part of superbuild, but the best way is to make a clean build of Seg3D (configure + build in a new binary directory).

File: src/Application/DatabaseManager/DatabaseManager.cc (hunk 8 above)
Comment (dzenanz) [reply src/Application/DatabaseManager/DatabaseManager.cc]: I just tried building and running. There are no SQL errors now like there were before.

**Comment by dzenanz** [comment]
I think this is ready. Should I beautify history by squashing and possibly reordering commits?

**Comment by dzenanz** [comment]
I created beautified commit history in #384. It is up to you whether to merge this (#382) or #384.

**Comment by allywarner** [comment]
@dzenanz thank you!

**Comment by RubioJr9** [comment]
I wiped my bin folder and that got rid of the issue I was having. However, I'm getting this error now:
```
/home/tpat/repos/seg3d/src/Core/ITKCommon/Optimizers/itkImageMosaicVarianceMetric.txx:435:59: error: cannot bind non-const lvalue reference of type 'itk::Transform<double, 2, 2>::JacobianPositionType&' {aka 'vnl_matrix_fixed<double, 2, 2>&'} to an rvalue of type 'itk::Transform<double, 2, 2>::JacobianPositionType' {aka 'vnl_matrix_fixed<double, 2, 2>'}
  435 |       t->ComputeJacobianWithRespectToPosition(point, data.J_);
```

**Comment by dzenanz** [comment]
@RubioJr9 which compiler/OS and their versions are you using?

**Comment by RubioJr9** [comment]
OS: Arch Linux(rolling release so no versioning)
Compiler: gcc 10.2.0

**Comment by dzenanz** [comment]
@RubioJr9 By reading [this SO question](https://stackoverflow.com/questions/18565167/non-const-lvalue-references), it seems that `data.J_` is not the same type which `ComputeJacobianWithRespectToPosition` expects, so implicit conversion is creating a temporary. I looked at the code and couldn't quite figure out why. Can you copy-paste the full error message with context (instantiation types etc)?

**Comment by RubioJr9** [comment]
I posted the log in the new PR. Let me know if you need more info.

**Pull request closed without merging**